# Segmenting and Clustering Neighborhoods in Toronto

### Assignment for the week 3 of the Coursera course"Applied Data Science Capstone"

Import requests package to get the wikipedia page and the BeautifulSoup module for scraping the web page

In [1]:
import requests
from bs4 import BeautifulSoup

Get the wikipedia page

In [2]:
link='https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
page = requests.get(link)

Transform the page in BeautifulSoup object for scraping

In [3]:
soup=BeautifulSoup(page.content,'html.parser')
#print(soup.prettify)

Get the table with the data

In [4]:
table=list(soup.find_all('table',class_='wikitable sortable'))[0]
#table

Get the list of the rows of the table

In [5]:
tr_list=table.find_all('tr')
r = tr_list.pop(0) # delete the firs element, containing the column labels
#tr_list

Bulild a list of dictionaries with the data following the requirenments of the asseignment

In [6]:
data=[]
pcode_prev = ''
for i, row in enumerate(tr_list):                  # iterate on the list of rows
    td_list = row.find_all('td')                   # build a list of the cells in the row
    borough = td_list[1].get_text()                # extract the value of borough in string format
    neighborhood = td_list[2].get_text().rstrip()  # extract the value of neighborhood in string format removing the "\n" characters
    if neighborhood == "Not assigned":
        neighborhood = borough
    if i == 0:                                     # to initialize the variables neighborhoods and borough_prev
        neighborhoods = neighborhood
        borough_prev = borough
    if( borough != "Not assigned"):
        pcode = td_list[0].get_text()        # current value of postal code
        if i == 0:
            data.append({'PostalCode': pcode,'Borough': borough,'Neighborhood': neighborhoods})
        else:
            if( pcode == pcode_prev):        # I'm assuming rows with the same postal code are in successive order 
                neighborhoods = neighborhoods + ', ' + neighborhood   # concatenate the neighborhood names
            else:                            # when find different postal code then append the row with the values of the previous postal code
                data.append({'PostalCode': pcode_prev,'Borough': borough_prev,'Neighborhood': neighborhoods})
                neighborhoods = neighborhood
        pcode_prev = pcode                  # register the previous vaule of postal code
        borough_prev = borough

Import pandas package

In [7]:
import pandas as pd

#### Build the dataframe from the list of dictionaries just built.

In [8]:
df=pd.DataFrame(data)
df=df[["PostalCode",'Borough','Neighborhood']]
df=df.drop(df.index[0])         # remove the first line
df = df.reset_index(drop=True)
df.head()

PostalCode           Borough                      Neighborhood
0        M3A        North York                         Parkwoods
1        M4A        North York                  Victoria Village
2        M5A  Downtown Toronto         Harbourfront, Regent Park
3        M6A        North York  Lawrence Heights, Lawrence Manor
4        M7A      Queen's Park                      Queen's Park

#### Dataframe shape:

In [9]:
df.shape

(102, 3)

There are 102 postal codes with assigned borough.

## Second part of the assignment

I have tried to use the Geocoder package but it doesn't work. So I download the csv file.

In [10]:
!wget -O toronto_geospatialdata.csv https://cocl.us/Geospatial_data

--2018-12-07 12:40:46--  https://cocl.us/Geospatial_data
Resolving cocl.us (cocl.us)... 159.8.72.228
Connecting to cocl.us (cocl.us)|159.8.72.228|:443... connected.
HTTP request sent, awaiting response... 301 Moved Permanently
Location: https://ibm.box.com/shared/static/9afzr83pps4pwf2smjjcf1y5mvgb18rr.csv [following]
--2018-12-07 12:40:47--  https://ibm.box.com/shared/static/9afzr83pps4pwf2smjjcf1y5mvgb18rr.csv
Resolving ibm.box.com (ibm.box.com)... 107.152.27.197
Connecting to ibm.box.com (ibm.box.com)|107.152.27.197|:443... connected.
HTTP request sent, awaiting response... 301 Moved Permanently
Location: https://ibm.ent.box.com/shared/static/9afzr83pps4pwf2smjjcf1y5mvgb18rr.csv [following]
--2018-12-07 12:40:47--  https://ibm.ent.box.com/shared/static/9afzr83pps4pwf2smjjcf1y5mvgb18rr.csv
Resolving ibm.ent.box.com (ibm.ent.box.com)... 107.152.26.211
Connecting to ibm.ent.box.com (ibm.ent.box.com)|107.152.26.211|:443... connected.
HTTP request sent, awaiting response... 302 Found
Loc

#### Load the data in a dataframe

In [11]:
toronto_geo_df = pd.read_csv('toronto_geospatialdata.csv')
toronto_geo_df.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

#### Make a join between the two dataframe on 'PostalCode' column

In [12]:
toronto_df = df
toronto_df = toronto_df.join(toronto_geo_df.set_index('Postal Code'),on='PostalCode')
toronto_df.head()

PostalCode           Borough                      Neighborhood   Latitude  \
0        M3A        North York                         Parkwoods  43.753259   
1        M4A        North York                  Victoria Village  43.725882   
2        M5A  Downtown Toronto         Harbourfront, Regent Park  43.654260   
3        M6A        North York  Lawrence Heights, Lawrence Manor  43.718518   
4        M7A      Queen's Park                      Queen's Park  43.662301   

   Longitude  
0 -79.329656  
1 -79.315572  
2 -79.360636  
3 -79.464763  
4 -79.389494

## Third part of the assignment

In [13]:
from geopy.geocoders import Nominatim

In [14]:
!conda install -c conda-forge folium=0.5.0 --yes
import folium

Fetching package metadata .............
Solving package specifications: .

# All requested packages already installed.
# packages in environment at /opt/conda/envs/DSX-Python35:
#
folium                    0.5.0                      py_0    conda-forge


#### Let's get the coordinates of Toronto

In [15]:
address = 'Toronto'

geolocator = Nominatim()
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.653963, -79.387207.


In [16]:
toronto_new_df = toronto_df[toronto_df['Borough'].str.contains('Toronto')]
toronto_new_df = toronto_new_df.reset_index()
print(toronto_new_df.shape)
toronto_new_df.head()

(38, 6)


index PostalCode           Borough               Neighborhood   Latitude  \
0      2        M5A  Downtown Toronto  Harbourfront, Regent Park  43.654260   
1      9        M5B  Downtown Toronto   Ryerson, Garden District  43.657162   
2     15        M5C  Downtown Toronto             St. James Town  43.651494   
3     19        M4E      East Toronto                The Beaches  43.676357   
4     20        M5E  Downtown Toronto                Berczy Park  43.644771   

   Longitude  
0 -79.360636  
1 -79.378937  
2 -79.375418  
3 -79.293031  
4 -79.373306

In [17]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=12)

# add markers to map
for lat, lng, borough, neighborhood in zip(toronto_new_df['Latitude'], toronto_new_df['Longitude'], toronto_new_df['Borough'], toronto_new_df['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

#### Define Foursquare Credentials and Version

In [18]:
# The code was removed by Watson Studio for sharing.

Your credentails:
CLIENT_ID: BL52B0FLGWV4JW4B0RFCZ1MDMY3CMEAAWDZJIR2JJCVJFGG3
CLIENT_SECRET:MELYILGOD5JEBOJ0RAGA2AAW3OXUDJIKLTWTCMZOF1GPL5A3


In [19]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [20]:
import json                               # library to handle JSON files
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

### Explore neighborhoods in Toronto
Let's create a function to loop the exploration process to all the neighborhoods in Toronto

In [21]:
LIMIT=100
radius=500

In [22]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    count=0
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        count += 1
        print(count, name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [23]:
toronto_venues = getNearbyVenues(toronto_new_df['Neighborhood'],toronto_new_df['Latitude'],toronto_new_df['Longitude'])

1 Harbourfront, Regent Park
2 Ryerson, Garden District
3 St. James Town
4 The Beaches
5 Berczy Park
6 Central Bay Street
7 Christie
8 Adelaide, King, Richmond
9 Dovercourt Village, Dufferin
10 Harbourfront East, Toronto Islands, Union Station
11 Little Portugal, Trinity
12 The Danforth West, Riverdale
13 Design Exchange, Toronto Dominion Centre
14 Brockton, Exhibition Place, Parkdale Village
15 The Beaches West, India Bazaar
16 Commerce Court, Victoria Hotel
17 Studio District
18 Lawrence Park
19 Roselawn
20 Davisville North
21 Forest Hill North, Forest Hill West
22 High Park, The Junction South
23 North Toronto West
24 The Annex, North Midtown, Yorkville
25 Parkdale, Roncesvalles
26 Davisville
27 Harbord, University of Toronto
28 Runnymede, Swansea
29 Moore Park, Summerhill East
30 Chinatown, Grange Park, Kensington Market
31 Deer Park, Forest Hill SE, Rathnelly, South Hill, Summerhill West
32 CN Tower, Bathurst Quay, Island airport, Harbourfront West, King and Spadina, Railway Lands,

In [24]:
print(toronto_venues.shape)
toronto_venues.head()

(1695, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Harbourfront, Regent Park               43.65426              -79.360636   
1  Harbourfront, Regent Park               43.65426              -79.360636   
2  Harbourfront, Regent Park               43.65426              -79.360636   
3  Harbourfront, Regent Park               43.65426              -79.360636   
4  Harbourfront, Regent Park               43.65426              -79.360636   

                 Venue  Venue Latitude  Venue Longitude        Venue Category  
0     Roselle Desserts       43.653447       -79.362017                Bakery  
1        Tandem Coffee       43.653559       -79.361809           Coffee Shop  
2      Cooper Koo YMCA       43.653191       -79.357947  Gym / Fitness Center  
3   Morning Glory Cafe       43.653947       -79.361149        Breakfast Spot  
4  Body Blitz Spa East       43.654735       -79.359874                   Spa

In [25]:
print(toronto_venues.groupby('Neighborhood').count().shape)
toronto_venues.groupby('Neighborhood').count().head(10)

(38, 6)


Neighborhood Latitude  \
Neighborhood                                                                
Adelaide, King, Richmond                                              100   
Berczy Park                                                            54   
Brockton, Exhibition Place, Parkdale Village                           21   
Business reply mail Processing Centre969 Eastern                       17   
CN Tower, Bathurst Quay, Island airport, Harbou...                     14   
Cabbagetown, St. James Town                                            45   
Central Bay Street                                                     84   
Chinatown, Grange Park, Kensington Market                              98   
Christie                                                               16   
Church and Wellesley                                                   85   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Adelaide, King, Richmond                                               100   
Berczy Park                                                             54   
Brockton, Exhibition Place, Parkdale Village                            21   
Business reply mail Processing Centre969 Eastern                        17   
CN Tower, Bathurst Quay, Island airport, Harbou...                      14   
Cabbagetown, St. James Town                                             45   
Central Bay Street                                                      84   
Chinatown, Grange Park, Kensington Market                               98   
Christie                                                                16   
Church and Wellesley                                                    85   

                                                    Venue  Venue Latitude  \
Neighborhood                                                                
Adelaide, King, Richmond                              100             100   
Berczy Park                                            54              54   
Brockton, Exhibition Place, Parkdale Village           21              21   
Business reply mail Processing Centre969 Eastern       17              17   
CN Tower, Bathurst Quay, Island airport, Harbou...     14              14   
Cabbagetown, St. James Town                            45              45   
Central Bay Street                                     84              84   
Chinatown, Grange Park, Kensington Market              98              98   
Christie                                               16              16   
Church and Wellesley                                   85              85   

                                                    Venue Longitude  \
Neighborhood                                                          
Adelaide, King, Richmond                                        100   
Berczy Park                                                      54   
Brockton, Exhibition Place, Parkdale Village                     21   
Business reply mail Processing Centre969 Eastern                 17   
CN Tower, Bathurst Quay, Island airport, Harbou...               14   
Cabbagetown, St. James Town                                      45   
Central Bay Street                                               84   
Chinatown, Grange Park, Kensington Market                        98   
Christie                                                         16   
Church and Wellesley                                             85   

                                                    Venue Category  
Neighborhood                                                        
Adelaide, King, Richmond                                       100  
Berczy Park                                                     54  
Brockton, Exhibition Place, Parkdale Village                    21  
Business reply mail Processing Centre969 Eastern                17 

In [26]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Yoga Studio  Adult Boutique  Afghan Restaurant  Airport  \
0            0               0                  0        0   
1            0               0                  0        0   
2            0               0                  0        0   
3            0               0                  0        0   
4            0               0                  0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant      ...        Toy / Game Store  \
0                 0                    0      ...                       0   
1                 0                    0      ...                       0   
2                 0                    0      ...                       0   
3                 0                    0      ...                       0   
4                 0                    0      ...                       0   

   Trail  Train Station  Vegetarian / Vegan Restaurant  Video Game Store  \
0      0              0                              0                 0   
1      0              0                              0                 0   
2      0              0                              0                 0   
3      0              0                              0                 0   
4      0              0                              0                 0   

   Vietnamese Restaurant  Wine Bar  Wine Shop  Wings Joint  Women's Store  
0                      0         0          0            0              0  
1                      0         0          0            0              0  
2                      0         0          0            0              0  
3                      0         0          0            0              0  
4                      0         0          0            0              0  

[5 rows x 230 columns]

In [27]:
toronto_onehot.shape

(1695, 230)

In [28]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped.head()

Neighborhood  Yoga Studio  \
0                           Adelaide, King, Richmond     0.000000   
1                                        Berczy Park     0.000000   
2       Brockton, Exhibition Place, Parkdale Village     0.000000   
3   Business reply mail Processing Centre969 Eastern     0.058824   
4  CN Tower, Bathurst Quay, Island airport, Harbo...     0.000000   

   Adult Boutique  Afghan Restaurant   Airport  Airport Food Court  \
0             0.0                0.0  0.000000            0.000000   
1             0.0                0.0  0.000000            0.000000   
2             0.0                0.0  0.000000            0.000000   
3             0.0                0.0  0.000000            0.000000   
4             0.0                0.0  0.071429            0.071429   

   Airport Gate  Airport Lounge  Airport Service  Airport Terminal  \
0      0.000000        0.000000         0.000000          0.000000   
1      0.000000        0.000000         0.000000          0.000000   
2      0.000000        0.000000         0.000000          0.000000   
3      0.000000        0.000000         0.000000          0.000000   
4      0.071429        0.142857         0.142857          0.142857   

       ...        Toy / Game Store  Trail  Train Station  \
0      ...                     0.0    0.0            0.0   
1      ...                     0.0    0.0            0.0   
2      ...                     0.0    0.0            0.0   
3      ...                     0.0    0.0            0.0   
4      ...                     0.0    0.0            0.0   

   Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                           0.01               0.0                    0.0   
1                           0.00               0.0                    0.0   
2                           0.00               0.0                    0.0   
3                           0.00               0.0                    0.0   
4                           0.00               0.0                    0.0   

   Wine Bar  Wine Shop  Wings Joint  Women's Store  
0      0.01        0.0          0.0           0.01  
1      0.00        0.0          0.0           0.00  
2      0.00        0.0          0.0           0.00  
3      0.00        0.0          0.0           0.00  
4      0.00        0.0          0.0           0.00  

[5 rows x 230 columns]

In [29]:
toronto_grouped.shape

(38, 230)

In [30]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [31]:
import numpy as np

In [32]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                           Adelaide, King, Richmond           Coffee Shop   
1                                        Berczy Park           Coffee Shop   
2       Brockton, Exhibition Place, Parkdale Village           Coffee Shop   
3   Business reply mail Processing Centre969 Eastern    Light Rail Station   
4  CN Tower, Bathurst Quay, Island airport, Harbo...        Airport Lounge   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Café            Steakhouse   American Restaurant   
1            Restaurant          Cocktail Bar           Cheese Shop   
2                  Café        Breakfast Spot             Pet Store   
3         Auto Workshop            Comic Shop           Pizza Place   
4      Airport Terminal       Airport Service       Harbor / Marina   

  5th Most Common Venue 6th Most Common Venue   7th Most Common Venue  \
0       Thai Restaurant        Clothing Store                   Hotel   
1                   Pub        Farmers Market      Seafood Restaurant   
2               Stadium                   Gym  Furniture / Home Store   
3         Burrito Place            Restaurant                 Brewery   
4      Sculpture Garden              Boutique                   Plane   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                   Bar                   Gym           Burger Joint  
1                  Café                Bakery             Steakhouse  
2    Italian Restaurant    Falafel Restaurant      Convenience Store  
3            Skate Park            Smoke Shop                    Spa  
4         Boat or Ferry          Airport Gate                Airport

Cluster neighbohoods with k-means

In [33]:
# import k-means from clustering
from sklearn.cluster import KMeans

In [34]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=int32)

In [35]:
toronto_merged = toronto_df

In [36]:
toronto_merged.shape

(102, 5)

In [37]:
toronto_merged.head()

PostalCode           Borough                      Neighborhood   Latitude  \
0        M3A        North York                         Parkwoods  43.753259   
1        M4A        North York                  Victoria Village  43.725882   
2        M5A  Downtown Toronto         Harbourfront, Regent Park  43.654260   
3        M6A        North York  Lawrence Heights, Lawrence Manor  43.718518   
4        M7A      Queen's Park                      Queen's Park  43.662301   

   Longitude  
0 -79.329656  
1 -79.315572  
2 -79.360636  
3 -79.464763  
4 -79.389494

In [38]:
kmeans.labels_.shape

(38,)

In [39]:
toronto_merged = toronto_new_df

# add clustering labels
toronto_merged['Cluster Labels'] = kmeans.labels_

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head() # check the last columns!

index PostalCode           Borough               Neighborhood   Latitude  \
0      2        M5A  Downtown Toronto  Harbourfront, Regent Park  43.654260   
1      9        M5B  Downtown Toronto   Ryerson, Garden District  43.657162   
2     15        M5C  Downtown Toronto             St. James Town  43.651494   
3     19        M4E      East Toronto                The Beaches  43.676357   
4     20        M5E  Downtown Toronto                Berczy Park  43.644771   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.360636               0           Coffee Shop                   Pub   
1 -79.378937               0           Coffee Shop        Clothing Store   
2 -79.375418               0           Coffee Shop                  Café   
3 -79.293031               0           Coffee Shop                 Trail   
4 -79.373306               0           Coffee Shop            Restaurant   

  3rd Most Common Venue 4th Most Common Venue      5th Most Common Venue  \
0                Bakery                  Café                       Park   
1                  Café        Cosmetics Shop  Middle Eastern Restaurant   
2            Restaurant                 Hotel             Cosmetics Shop   
3                   Pub      Department Store       Ethiopian Restaurant   
4          Cocktail Bar           Cheese Shop                        Pub   

  6th Most Common Venue        7th Most Common Venue 8th Most Common Venue  \
0    Mexican Restaurant                      Theater        Breakfast Spot   
1   Japanese Restaurant                  Pizza Place   American Restaurant   
2             Gastropub                       Bakery                  Park   
3     Electronics Store  Eastern European Restaurant   Dumpling Restaurant   
4        Farmers Market           Seafood Restaurant                  Café   

  9th Most Common Venue 10th Most Common Venue  
0            Restaurant         Farmers Market  
1    Italian Restaurant             Restaurant  
2    Italian Restaurant         Clothing Store  
3            Donut Shop       Doner Restaurant  
4                Bakery             Steakhouse

In [41]:
# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=12)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i+x+(i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters